### ISSUE TITLE
Galaxy locations are inconsistent

### ISSUE THREAD

**Issue by bhilbert4** [spacetelescope/mirage#189]
As I was making a dataset to be used for an image registration hack day on Monday, I found something strange with the positions of galaxies in the seed images. I was simulating 6 pointings for all 10 detectors of NIRCam. As I blinked the seed images for all of the NRCA1 files, I saw that the stars looked as expected from image to image. All were in the same place on the sky in all images. But the galaxies seem to jump around, and some are present in one pointing but not the others. Note that I'm using the same galaxy catalog for all pointings.

I wonder if the filter looking for galaxies outside the fov is bad. OR...I saw some lines in the code earlier that implied that if the core of the galaxy was outside the fov, that the galaxy was skipped. That's not quite right, since you could have a galaxy extending onto the detector from just outside.

I don't have time to fix this bug for hack day. I'll just re-create the data with stars only. But this is a bug that should be investigated soon.

Data are in central store (grp) in wit/nircam/hilbert/simulated_ramps/image_reg_hack_day

**Comment by bhilbert4** [comment]
So far, it looks like this is an issue only when using the new webbpsf libraries with the spatially varying PSFs. I ran my test case using the current master and there was no problem. I ran 3 dithers of one detector, and found the same 4 or 5 galaxies in all three, moving as expected between them.

With the new webbpsf libraries, Mirage is finding 19 or 20 galaxies per image. Some move as expected from dither to dither, and others don't.

**Comment by bhilbert4** [comment]
This is very confusing, since the PSF isn't convolved with the galaxy stamp image until after sources off the chip have been filtered out.

**Comment by bhilbert4** [comment]
I think this is a problem with the use of pysiaf vs distortion reference files to get information. It's a problem here because the distortion reference files being used by Mirage are still the old versions that contain the round trip error, whereas pysiaf does not. 

The local roll angle and the attitude matrix are being created using pysiaf, but coordinate transforms default to come from the reference file. So we're mixing sources when we shouldn't be. 

Quick fix, which I will test now, is to set the distortion reference files to None, so that pysiaf will be used for everything.

More permanent fix will be to make sure consistent sources are being used for all siaf-info.

**Comment by bhilbert4** [comment]
The problem appears to be a little more complex than described above. 
I think this is due to the fact that the ePSF branch currently does not have the source filtering implemented where we throw out sources very far (>4096 pixels) from the detector. Using one of the disappearing galaxies as a test, I found that the galaxy's RA and Dec were in fact very far (~700 arcseconds) from the detector. Running it's RA and Dec through pysiaf's coordinate transform (using only pysiaf) I found that the calculations were putting the galaxy on the detector, only 16 arcseconds from the center. My theory is that this far from the detector, the distortion model (since it is a polynomial) can be doing just about anything and it's not surprising that the output is wildly incorrect.

So hopefully this bug can be fixed simply by rebasing the ePSF branch, since the source filtering is already in the master branch.

**Comment by bhilbert4** [comment]
Rebasing, which added the source filtering to the ePSF branch, seems to have solved most of the problem. There are no more disappearing galaxies in the resulting seed images. The number of point sources in each detector's field of view is still not quite the same as when using the master branch. I believe this is because of the difference in the distortion model between the old version with the round trip error, and what's in the yet-to-be-delivered pysiaf version. 

Currently, pysiaf is used to calculate the local roll angle and the attitude matrix, (v2ref and v3ref are the primary inputs), but if a distortion reference file is provided, then the coefficients in that file are used for the distortion model. So by specifying the old reference file in this simulation dataset, the old model coefficients are being used with the new v2ref, v3ref (from pysiaf), resulting in an incorrect model. v2ref, v3ref have both shifted between the old and new models. So I need to re-run mirage using new distortion reference files (or no reference files, in which case Mirage falls back to using pysiaf for everything).

**Comment by bhilbert4** [comment]
Complicating this investigation even further: It seems that when running on witserv, that only the master branch is used. I had pulled down the EPSF branch to my copy of the repo on witserv, and checked out that branch. I then ran mirage. After some investigation though, it appears that witserv is still running the mirage from the master branch. On my laptop, whichever branch is checked out is the one that is run.

This may be related to the fact that the version of git on witserv is from 2012. I put in a ticket to IT a couple weeks ago to get it updated.
